### ISSUE TITLE
[2019] Parameter-Efficient Transfer Learning for NLP

### ISSUE THREAD

**Issue by cfiken** [cfiken/paper-reading#127]
## 0. 論文

| タイトル | Parameter-Efficient Transfer Learning for NLP |
|:-------|:--|
| リンク  | https://arxiv.org/abs/1902.00751 |
| 著者    | Neil Houlsby, Andrei Giurgiu, Stanislaw Jastrzebski, Bruna Morrone, Quentin de Laroussilhe, Andrea Gesmundo, Mona Attariyan, Sylvain Gelly (Google Research + Jagiellonian University) |
| 投稿日付 | ICML 2019 |

## 1. どんなもの？

NLP において、大規模な事前学習モデルを finetuning することが下位タスクを解くのに効果的である一方で、事前学習モデルを finetuning したパラメータ全てがタスクごとに必要になるためパラメータ数に対しての効率が悪い。
提案手法である Adapter は、事前学習モデル部分は固定し、モデルに少量のレイヤを追加してその部分を下位タスクで学習することで少ないパラメータ数で多くのタスクに対応できる。
GLUE で BERT ベースの finetuning とそこまで変わらない (0.4%落ち) レベルの性能を、パラメータ数において finetuning では 9 x (BERT のパラメータ数) が必要なのに対して、1.3 x (BERT のパラメータ数) で達成。

## 2. 先行研究と比べてどこがすごい？

事前学習モデルから下位タスクへの適用方法として新しい手法である Adapter を Transformer に組み込んだ。モデルの途中にレイヤを追加し、事前学習モデル部分は固定したまま新しいレイヤのみを学習する。
多くのタスクを解く場合に、必要なパラメータ数を圧倒的に減らした状態で finetuning に近いスコアを達成。

## 3. 技術や手法のキモはどこ？

<img width="616" alt="スクリーンショット 2021-04-16 16 35 38" src="https://user-images.githubusercontent.com/1255485/114988907-d2fbc180-9ed1-11eb-8708-1211f705910d.png">

- モデル
    - 事前学習モデル (論文では BERT) を用意する
    - モデルの各 Transformer layer の attention block, ffn block の後に上記画像のように Adapter 構造を取り入れる
    - Adapter 内は skip connection に加えて、パラメータ削減のため bottleneck 構造にしている
    - Adapter のパラメータの初期値は、事前学習モデルと変わらないように near-zero (skip connection だけがほぼ残る)で行う
- 学習
    - 事前学習モデル部分のパラメータは全て固定する
    - Adapter 部分を下位タスクで学習
        - パラメータの初期値が Adapter 部分が identity に近いところから始まるので、調整の必要のないタスクでは何もしない Adapter ができる
        - Adapter 部分に加えて Layer Normalization のパラメータも一緒に学習している

## 4. どうやって有効だと検証した？

BERT に対して、finetuning, variable finetuning (finetuning 時に一部レイヤを固定する), 提案手法である Adapter の3つの学習手法で、GLUE とその他分類タスクを多数用意して実験。
全体として、複数タスクにおける実質パラメータ数が大きく削減する一方、精度は finetuning に比べて少しだけ落ちる程度を達成。
分類タスク以外で SQuAD でも同様の結果に。
別の研究で行われている、Layer Normalization のパラメータのみを tune する手法でも試しており、Adapter の改善を確認している。

<img width="1170" alt="スクリーンショット 2021-04-16 16 48 18" src="https://user-images.githubusercontent.com/1255485/114990625-a9439a00-9ed3-11eb-9a14-f084ae5e7962.png">

学習したモデルの分析として、下位タスクで学習してから Adapter レイヤを抜いた場合の性能の変化から、1つのレイヤごとの影響は小さいが全体では大きいこと、下位レイヤより上位レイヤのほうが影響が大きいことなどを確認 (これは他研究や finetuning での知見と一致する)。

他にも多数の工夫を行っているが、改善にはつながらなかったとのこと。

## 5. 議論はある？

- [IMO] 適用可能なのはそうとして、実用は大変そう。タスクごとに別でパラメータを保存して、事前学習モデルに組み込む(外から渡す?)形で使うことになるので、推論速度にも影響ありそう

## 6. 次に読むべき論文は？

- Efficient parametrization of multi-domain deep neural networks
    - https://arxiv.org/abs/1803.10082
    - Sylvestre-Alvise Rebuffi, Hakan Bilen, Andrea Vedaldi
    - Adapter のアイディアの元論文
